app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/vpn/service/VpnActiveNetworkManager.kt

```diff
@@ -0,0 +1,206 @@
+    private val vpnFeaturesRegistry: VpnFeaturesRegistry,
+    @AppCoroutineScope private val coroutineScope: CoroutineScope,
+    private val context: Context,
+) : VpnServiceCallbacks {
+
+    private val isInterceptDnsRequestsEnabled: Boolean
+        get() = appTpFeatureConfig.isEnabled(AppTpSetting.InterceptDnsRequests)
+
+    private var currentNetwork: AtomicReference<String?> = AtomicReference(null)
+
+    private val cellularNetworkRequest = NetworkRequest.Builder()
+        .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
+        .build()
+
+    private val wifiNetworkRequest = NetworkRequest.Builder()
+        .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
+        .build()
+
+    override fun onVpnStarted(coroutineScope: CoroutineScope) {
+        if (runBlocking { !shouldHandleNetworkChanges() }) {
+            logcat { "Do not handle network changes" }
+            return
+        }
+
+        if (isInterceptDnsRequestsEnabled) {
+            service.get().configureUnderlyingNetworks()
+
+            (context.applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager?)?.let {
+                it.safeRegisterNetworkCallback(wifiNetworkRequest, wifiNetworkCallback)
+                it.safeRegisterNetworkCallback(cellularNetworkRequest, cellularNetworkCallback)
+            }
+        } else {
+            logcat { "DNS based blocking is disabled, noop" }
+        }
+    }
+
+    override fun onVpnReconfigured(coroutineScope: CoroutineScope) {
+        // both when VPN is first started and reconfigured we check if we need to handle network changes
+        onVpnStarted(coroutineScope)
+    }
+
+    override fun onVpnStopped(
+        coroutineScope: CoroutineScope,
+        vpnStopReason: VpnStopReason,
+    ) {
+        (context.applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager?)?.let {
+            it.safeUnregisterNetworkCallback(wifiNetworkCallback)
+            it.safeUnregisterNetworkCallback(cellularNetworkCallback)
+        }
+    }
+
+    private suspend fun restartIfActiveNetworkChanged() {
+        val service = service.get()
+        if (ignoreRestartEvent()) {
+            service.configureUnderlyingNetworks()
+            logcat { "AppTP (only) is not enabled, ignoring restartIfActiveNetworkChanged() event" }
+            return
+        }
+
+        val activeNetwork = service.getActiveNetwork()
+        if (activeNetwork == null) {
+            logcat { "Current active network is NULL, ignore restart" }
+            return
+        }
+        val previousNetwork = currentNetwork.getAndSet(activeNetwork.toString())
+        if (previousNetwork == null) {
+            logcat { "Previous network is NULL, ignore restart" }
+            return
+        }
+        if (currentNetwork.get() == null) {
+            logcat { "Current network is NULL, ignore restart" }
+            return
+        }
```

Comment:
I think we don't need this anymore

Reply:
Yeah, it feels odd both ways because the it's type is `AtomicReference<String?>` so to have a `null` initial value.
I think I am gonna leave it as a paranoid check, if you don't feel strongly about it